### PR TITLE
[Fix] Dagger Methods in ControlBlocks and Scaleblock

### LINKED ISSUE #215: Dagger/adjoint operation of qd.MCRX/qd.MCRY/qd.MCRZ functions
Hi Qadence team,

display(qd.MCRX([1,2,3],0,pi/2))

display(qd.MCRX([1,2,3],0,pi/2).dagger())
TypeError: MCRX.__init__() takes 4 positional arguments but 6 were given

Please also check
display(qd.MCRY([1,2,3],0,pi/2).dagger())
display(qd.MCRZ([1,2,3],0,pi/2).dagger())

### PR DESCRIPTION
**Pull request by vincentelfving**
fixes #215 

TBD:

- [x] unit tests to verify this
- [x] also fix other controlled and multi controlled operations, for their dagger methods.

_Edit_:
fixes https://github.com/pasqal-io/qadence/issues/217

### CODE REVIEW

**Comment by dominikandreasseitz** [comment]
thanks @vincentelfving! lets add some tests, do you want me to take that over?

**Comment by Roland-djee** [comment]
@vincentelfving @dominikandreasseitz What is the status on this ?

**Comment by dominikandreasseitz** [comment]
i will finish this. sorry i didnt get to it last week

File: qadence/blocks/primitive.py [hunk 1]
```diff
@@ -200,11 +207,10 @@ def _from_dict(cls, d: dict) -> ParametricBlock:
...
         return cls(target, params)  # type: ignore[call-arg]
 
-    def dagger(self) -> ParametricBlock:  # type: ignore[override]
+    def dagger(self) -> ParametricBlock:
         exprs = self.parameters.expressions()
-        args = tuple(-extract_original_param_entry(param) for param in exprs)
-        args = args if -1 in self.qubit_support else (*self.qubit_support, *args)
-        return self.__class__(*args)  # type: ignore[arg-type]
+        params = tuple(-extract_original_param_entry(param) for param in exprs)
+        return type(self)(*self.qubit_support, *params)  # type: ignore[arg-type]
```
Comment (jpmoutinho) [inline qadence/blocks/primitive.py:213]: I know this is just implementing the same logic that was there before, but this assumes that all `ParametricBlocks` are implemented as `exp(i * param * some_block)`. I guess that is currently true right? Just something to keep in mind, I don't think I have a better solution.

File: qadence/operations.py [hunk 2]
```diff
@@ -387,7 +369,7 @@ def __pow__(self, power: int) -> AbstractBlock:
         return self
 
     def dagger(self) -> Zero:
-        return Zero()
+        return self
```
Comment (jpmoutinho) [inline qadence/operations.py:372]: Do you need to override the method from the `PrimitiveBlock` here?

**Review by jpmoutinho — APPROVED** [review]
Nice 👍

File: qadence/blocks/primitive.py (hunk 1 above)
Comment (dominikandreasseitz) [reply qadence/blocks/primitive.py:213]: yeah, the idea is that the "special" cases just override this method

File: qadence/operations.py (hunk 2 above)
Comment (dominikandreasseitz) [reply qadence/operations.py:372]: you mean putting the override decorator?

Comment (dominikandreasseitz) [reply qadence/operations.py:372]: yes

Comment (dominikandreasseitz) [reply qadence/operations.py]: done

**Review by Roland-djee — APPROVED** [review]

**Pull request merged**